generate/post_process_json.go

```diff
@@ -0,0 +1,113 @@
+package generate
+
+import (
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"reflect"
+	"strings"
+)
+
+// jsonNameFromProtobufTag parse json name from protobuf tag, returns empty string if no json name is defined.
+// example: protobuf:"bytes,1,opt,name=FirstName,json=first_name,proto3"
+// result: first_name
+func jsonNameFromProtobufTag(tag string) string {
+	const jsonPartName = "json="
+	if len(tag) == 0 {
+		return ""
+	}
+	parts := strings.Split(tag, ",")
+	for _, part := range parts {
+		if strings.HasPrefix(part, jsonPartName) {
+			return strings.TrimPrefix(part, jsonPartName)
+		}
+	}
+	return ""
+}
+
+// jsonNameFromJSONTag returns the json name from json tag, returns empty string if not defined.
+func jsonNameFromJSONTag(tag string) string {
+	if len(tag) == 0 {
+		return ""
+	}
+	parts := strings.Split(tag, ",")
+	if len(parts) == 0 {
+		return ""
+	}
+	return parts[0]
+}
+
+type textArea struct {
+	pos  token.Pos
+	end  token.Pos
+	text string
+}
+
+func jsonTagPostProcessor(input []byte) ([]byte, error) {
+	const (
+		jsonTagName     = "json"
+		protobufTagName = "protobuf"
+	)
+	var textAreas []textArea
+
+	fset := token.NewFileSet()
+	f, err := parser.ParseFile(fset, "", input, parser.ParseComments)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, decl := range f.Decls {
+		// check if is generic declaration
+		genDecl, ok := decl.(*ast.GenDecl)
+		if !ok {
+			continue
+		}
+
+		var typeSpec *ast.TypeSpec
+		for _, spec := range genDecl.Specs {
+			if ts, tsOK := spec.(*ast.TypeSpec); tsOK {
+				typeSpec = ts
+				break
+			}
+		}
+
+		// skip if can't get type spec
+		if typeSpec == nil {
+			continue
+		}
+
+		// not a struct, skip
+		structDecl, ok := typeSpec.Type.(*ast.StructType)
+		if !ok {
+			continue
+		}
+
+		for _, field := range structDecl.Fields.List {
+			tagValue := field.Tag.Value
+			tagValue = strings.TrimLeft(tagValue, "`")
+			tagValue = strings.TrimRight(tagValue, "`")
+			tag := reflect.StructTag(tagValue)
+
+			jsonName := jsonNameFromJSONTag(tag.Get(jsonTagName))
+			protobufJSONName := jsonNameFromProtobufTag(tag.Get(protobufTagName))
+
+			if jsonName != "" && protobufJSONName != "" && jsonName != protobufJSONName {
+				text := strings.Replace(tagValue, `json:"`+jsonName, `json:"`+protobufJSONName, 1)
+				textAreas = append(textAreas, textArea{
+					pos:  field.Tag.Pos(),
+					end:  field.Tag.End() - 1,
+					text: text,
+				})
+			}
+		}
+	}
+
+	output := make([]byte, len(input))
+	copy(output, input)
+	for i := range textAreas {
+		area := textAreas[len(textAreas)-i-1]
+		output = append(output[:area.pos], append([]byte(area.text), output[area.end-1:]...)...)
+	}
```

Comment:
This seems fine, but it's a bit weird. It would be simpler to modify the tags in-place in the AST, and then print out the file again. For example, it would be nearly impossible to make "off by one" errors with that method, and it would be much harder to end up with invalid syntax.

Reply:
Also note that with `ast.Inspect` you can modify nodes in-place, but you can't easily replace them with a different node altogether. You could use https://godoc.org/golang.org/x/tools/go/ast/astutil#Apply for that. I don't think you need to replace entire nodes here, but FYI in case you need to in the future.

Reply:
Updated rewriting the tag directly in ast.File.